### PR TITLE
Site Editor Tracking: Track browsing sidebar actions

### PR DESCRIPTION
**Pull request by david-szabo97**
#### Changes proposed in this Pull Request

* Track "Back to dashboard button click" in browsing sidebar `wpcom_block_editor_close_click`
* Track "Browsing Sidebar Open" action `wpcom_block_editor_nav_sidebar_open`
* Track "Template creation via browsing sidebar" `wpcom_block_editor_nav_sidebar_item_add`
  * Includes two event properties:
    * `item_type`: This is always `template`, indicating a template is being created
    * `item_slug`: Slug of the template
* Track "Template creation via browsing sidebar" `wpcom_block_editor_nav_sidebar_item_edit`
  * Includes two event properties:
    * `item_type`: This is either `template` or `template_part`, indicating what type of entity being switched to
    * `item_id`: ID of the template / template part
    * `item_slug`: Slug of the template / template part

#### Testing instructions

* Follow steps to enable tracking debugging at PCYsg-nrf-p2
* Open Site Editor

1. **Open the browsing sidebar (W/your site logo button in the top left corner**
1.1 Confirm `wpcom_block_editor_nav_sidebar_open` is triggered

2. **In the Template submenu create a new "Search" template (or anything else that is available)**
2.1 Confirm `wpcom_block_editor_nav_sidebar_item_add` is triggered with `item_type = "template"` and `item_slug = "search"`

3. **Open the browsing sidebar again and click on Templates -> Index**
3.1 Confirm `wpcom_block_editor_nav_sidebar_item_edit` is triggered with `item_type = "template"`, `item_id = "YOUR_THEME//index"` and `item_slug = "index"`

4. **Open the browsing sidebar again and click on Template Parts -> Headers -> Header**
4.1 Confirm `wpcom_block_editor_nav_sidebar_item_edit` is triggered with `item_type = "template_part"` and `item_slug = "header"`

5. **Open the browsing sidebar and click on the "Back to dashboard" button**
5.1 Confirm `wpcom_block_editor_close_click` is triggered

Related to #53410

### CODE REVIEW

**Comment by david-szabo97** [comment]
I'm not sure about the event names here. Should we stick to `calypso_editor_sidebar_*` ? 🤔

**Comment by Addison-Stavlo** [comment]
> I'm not sure about the event names here. Should we stick to calypso_editor_sidebar_* ?

I would assume we should leave the `calypso_*` name for the dotcom version of the sidebar that is live in the post editor.  Since this corresponds to the core sidebar I assume it should have separate event names? (cc @kurt213 and @ianstewart )

Perhaps `wpcom_block_editor_core_nav_sidebar_*` or something similar? (😆  these names seem like they could get a little lengthy)

**Comment by ianstewart** [comment]
> I would assume we should leave the calypso_* name for the dotcom version of the sidebar that is live in the post editor. Since this corresponds to the core sidebar I assume it should have separate event names

This seems logical to me too.

> Perhaps wpcom_block_editor_core_nav_sidebar_* or something similar? (😆 these names seem like they could get a little lengthy)

Agreed on the naming getting lengthy. :) I have no other strong opinion on the naming but that. :)

**Comment by david-szabo97** [comment]
`wpcom_block_editor_core_nav_sidebar_*` is lengthy but seems appropriate :D Renamed the events. If got any idea to make it shorter then I'm happy to rename then again.

**Review by Addison-Stavlo — COMMENTED** [review]
This is looking good and behaving well.

Do you think we should add similar tracking for `setPage` ?  So `wpcom_block_editor_core_nav_sidebar_item_edit` would work for the posts/pages under the content section of the menu as well?

**Comment by ianstewart** [comment]
> If got any idea to make it shorter then I'm happy to rename then again.

I think as long as we document all of these event names we're good.

**Review by Copons — APPROVED** [review]
I'm approving this because looks and works good, and I'm personally OK with long event names. ✨ 

I've got a question (not a blocker) about clicking the WP icon.
We are tracking both `wpcom_block_editor_close_click` and `wpcom_block_editor_core_nav_sidebar_open`.
It's worth noting that, in the Post Editor, we only track `close_click`, even though in that case too it opens a (different, Dotcom-specific) navigation sidebar.
I would assume that for the Post Editor we are inferring that `close_click` also opens the sidebar. Should we infer the same for the Site Editor instead of explicitly tracking the event? 🤔

**Comment by Addison-Stavlo** [comment]
> I would assume that for the Post Editor we are inferring that close_click also opens the sidebar. Should we infer the same for the Site Editor instead of explicitly tracking the event?

My assumption is that the `*_close_click` action was added long ago, before we had any nav sidebars and the W icon closed the editor instead of opening a panel.  It doesn't really make sense where it is anymore, as you noted it does not fire when the editor is closed but when the nav panel is opened.  If anything you think it would be best moved to the `< Dashboard` button inside the nav panel.

For the Site Editor - I wonder if we should remove this altogether or update `< Dashboard` button to use this name?  Since we are tracking the nav panel opening and the `< Dashboard`, it just seems like excess noise where it sits.  I also wonder if it should be left where it is for consistency with its odd behavior in the post editor.

For the Post Editor - This seems a little more tricky given that it has been in place with this odd behavior since the dotcom nav panel was added months back.  As @Copons considered, it could now be used to determine whether the nav panel in the post editor is opened.  I fear that changing this event here could make it difficult to analyze data due to the discrepancy in context across time, although it definitely seems like it would be most semantic if it was actually on the close button inside the nav panel. 🤔 

cc @kurt213 as well if you have any thoughts regarding this.

**Comment by Copons** [comment]
Yeah I guess it might make sense to move `close_click` to the back buttons _inside_ the sidebars, for all editors that have a nav sidebar tied to the WP button.

| Post Editor | Site Editor |
| - | - |
| Screenshot 2021-06-18 at 13 59 11 | Screenshot 2021-06-18 at 13 59 31 |

For these editors, track `nav_sidebar_open` for the WP button click 
Also, at this point I would not define the (Site Editor's) sidebar with `core_`, and use the same event regardless. We can infer the type of the sidebar by the editor type property of the event.

On the other hand, editors that are closed by simply clicking the WP icon should keep the `close_click` event on it.
In other words: let's track the behaviour, not the UX.

As mentioned by @Addison-Stavlo, this would mean changing what `close_click`, which is a tracked event, is used for.
I'd wait for a @kurt213 confirmation before acting on it. 🙂

**Comment by kurt213** [comment]
Hi all,

apologies for the delay in getting back on this PR.

> I would assume we should leave the calypso_*
> Perhaps wpcom_block_editor_core_nav_sidebar_*

I think the only 'unofficial rule' I've heard around using `calypso_` vs. `wpcom_` is `calypso` is generally used for client side and `wpcom` for server side events, but no strong preference from my side for this to be upheld if it's not being used.

Happy with lengthier names. I'd generally lean to having a longer event name that is clear than shortening an event name that could become difficult to understand in the future.

> wpcom_block_editor_close_click

Interesting, I remember spotting this event when documenting the audit list and thought the behaviour was strange myself!

100% agree with @Copons to 'track the behaviour, not the UX.' Happy for the event to be moved accordingly to where it semantically should be. 

I'm happy to update the description in Tracks to mention the date this change went live and explain the change made just so we have it documented. The event isn't being used in any active Funnels so my assumption is it has not been used for any significant analysis where the historical discrepancy will cause an issue.

**Comment by david-szabo97** [comment]
* Removed `wpcom_block_editor_core_nav_sidebar_editor_close` event and replaced with `wpcom_block_editor_close_click`. When you click on the Back to dashboard button then it will track the latter.
* EDIT: Also removed core prefix.

**Comment by david-szabo97** [comment]
- Opened a [follow-up PR](https://github.com/Automattic/wp-calypso/pull/53864) according to the conversation above about close editor event tracking.
- E2E tests are implemented in the PR above. We can test it by checking out to the other PR and patching this one.

**Comment by Addison-Stavlo** [comment]
Related to my comment on the follow-up PR noted https://github.com/Automattic/wp-calypso/pull/53864#pullrequestreview-688467483

Im finding there are already a `calypso_editor_sidebar_editor_close` event set up in the editing toolkit inside the implementation of the dotcom sidebar.  Although the close version doesn't seem to work in all cases, I wonder if what we should be doing here.  We could:

* Fix the calypso sidebar close event so its always firing when the button is used, and add a similar event for this core sidebar here in wpcom-block-editor.  In this case I think it would make sense to limit `wpcom_block_editor_close_click` to the rare case there is no sidebar?
* Get rid of that calypso version close editor event and just use `wpcom_block_editor_close_click` in all cases?
* Keep events for both sidebar versions AND use `wpcom_block_editor_close_click` as well?

**Comment by kurt213** [comment]
> `calypso_editor_sidebar_editor_close`

From my understanding, I don't think there is any need for the separate event unless it's telling us a separate behaviour? We will just need to make sure we mark in Tracks that this event is deprecated and replaced by `wpcom_block_editor_close_click`.

I saw in the Tracks description the event  "_often takes the user back to the pages or posts list, but can also be overridden (using the filter) to take the user to the customer homepage or themes gallery instance._" Not sure if that's some unique functionality we need to take into consideration for `wpcom_block_editor_close_click`?

**Comment by Addison-Stavlo** [comment]
> Not sure if that's some unique functionality we need to take into consideration for wpcom_block_editor_close_click?

I don't think so, it sounds like the functionality for the close buttons to take the user back to the entry point they came from.  The site editor close button uses a slot/fill to enable this as well, although right now there aren't many entry points for it. 😁

File: test/e2e/specs/specs-wpcom/wp-calypso-gutenberg-site-editor-tracking-spec.js [hunk 1]
```diff
@@ -26,6 +26,94 @@ const host = dataHelper.getJetpackHost();
...
+		! ( await driverHelper.isElementLocated(
+			driver,
+			By.css( '.edit-site-navigation-panel__back-to-dashboard' )
+		) )
+	) {
+		await driverHelper.clickWhenClickable(
+			driver,
+			By.css( '.components-navigation__back-button' )
+		);
+	}
```
Comment (WunderBart) [inline test/e2e/specs/specs-wpcom/wp-calypso-gutenberg-site-editor-tracking-spec.js]: We should be able to use the native `wait` loop here:
```suggestion
	await this.driver.wait(
		async function ( driver ) {
			if (
				await driverHelper.isElementNotLocated(
					driver,
					By.css( '.edit-site-navigation-panel__back-to-dashboard' )
				)
			) {
				await driverHelper.clickWhenClickable(
					driver,
					By.css( '.components-navigation__back-button' )
				);
				return false;
			}
			return true;
		},
		config.get( 'explicitWaitMS' ),
		'Could not reach the "Dashboard" button'
	);
```

This makes sure [the condition is polled every 200ms](https://github.com/SeleniumHQ/selenium/blob/68f3edcb9b28925757d3f19ba2a255d7d2d6b5b6/javascript/node/selenium-webdriver/lib/webdriver.js#L807). It will also timeout at the provided value with a useful error message. The timeout value can be tweaked - the `explicitWaitMS` is 20sec AFAIR. If you think it should be more or less, please use the `explicitWaitMS` as a base value, i.e. `explicitWaitMS / 5` or `explicitWaitMS * 2`, etc.

**Review by WunderBart — CHANGES_REQUESTED** [review]
Looks good! Just one comment from me.

**Review by WunderBart — APPROVED** [review]
Looks like the linter check is failing. Otherwise, it looks good from the E2E perspective! 🚀

File: test/e2e/specs/specs-wpcom/wp-calypso-gutenberg-site-editor-tracking-spec.js [hunk 2]
```diff
@@ -634,8 +735,172 @@ describe( `[${ host }] Calypso Gutenberg Site Editor Tracking: (${ screenSize })
...
+					'Back'
+				);
+				await driverHelper.clickWhenClickable( this.driver, backButtonToTemplatesLocator2 );
+
+				const isBackToDashboardLocated = await driverHelper.isElementLocated(
+					this.driver,
+					By.css( '.edit-site-navigation-panel__back-to-dashboard' )
+				);
+
+				assert( isBackToDashboardLocated );
```
Comment (WunderBart) [inline test/e2e/specs/specs-wpcom/wp-calypso-gutenberg-site-editor-tracking-spec.js]: I think in this case it might be better to use `isElementEventuallyLocatedAndVisible`. It would provide the same result, but at the same time be less prone to flakiness as it would wait for the button in case it needs a sec to be mounted after the click:
```suggestion
				const isBackToDashboardLocated = await driverHelper.isElementEventuallyLocatedAndVisible(
					this.driver,
					By.css( '.edit-site-navigation-panel__back-to-dashboard' )
				);

				assert( isBackToDashboardLocated );
```

**Review by WunderBart — APPROVED** [review]
Just one comment, otherwise this one's good to go from the E2E perspective! Thanks! :shipit:

**Review by Addison-Stavlo — APPROVED** [review]
Tested again and its all working well!

As a follow up idea - Do you think it would be possible to add tracking for selecting items in the content section of the menu?
